src/pybind/mgr/dashboard_v2/controllers/filesystem.py

```diff
@@ -0,0 +1,357 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import collections
+from collections import defaultdict
+import json
+import time
+
+import cherrypy
+from mgr_module import CommandResult
+
+from ..tools import ApiController, AuthRequired, BaseController, NotificationQueue
+
+
+LOG_BUFFER_SIZE = 30
+
+
+@ApiController('filesystem')
+@AuthRequired()
+class Filesystem(BaseController):
+    def __init__(self):
+        super(Filesystem, self).__init__()
+
+        # Stateful instances of CephFSClients, hold cached results.  Key to
+        # dict is FSCID
+        self.cephfs_clients = {}
+
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def clients_data(self, fs_id):
+        try:
+            fs_id = int(fs_id)
+        except ValueError:
+            raise cherrypy.HTTPError(400,
+                                     "Invalid filesystem id {0}".format(fs_id))
+
+        return self._clients(fs_id)
+
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def filesystem_data(self, fs_id):
+        try:
+            fs_id = int(fs_id)
+        except ValueError:
+            raise cherrypy.HTTPError(400,
+                                     "Invalid filesystem id {0}".format(fs_id))
+
+        return self.fs_status(fs_id)
+
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def mds_counters(self, fs_id):
+        """
+        Result format: map of daemon name to map of counter to list of datapoints
```

Comment:
You can help our type checker by adding this to the docstring: `:rtype: dict[str, dict[str, list]]`

Reply:
done